store/models/job_spec.go

```diff
@@ -339,3 +340,77 @@ func NormalizeSpecJSON(s string) (string, error) {
 
 	return utils.NormalizedJSONString(js)
 }
+
+// ServiceAgreement connects job specifications with on-chain encumbrances.
+type ServiceAgreement struct {
+	Encumbrance Encumbrance `json:"encumbrance" storm:"inline"`
+	ID          string      `json:"id" storm:"id,unique"`
+	JobSpecID   string      `json:"jobSpecID"`
+	jobSpec     JobSpec
+}
+
+// NewServiceAgreementFromRequest builds a new ServiceAgreement.
+func NewServiceAgreementFromRequest(sar ServiceAgreementRequest) (ServiceAgreement, error) {
+	sa := ServiceAgreement{}
+
+	sa.Encumbrance = sar.Encumbrance
+	sa.jobSpec = sar.JobSpec
+
+	b, err := utils.HexToBytes(sa.Encumbrance.ABI(), sar.Digest)
+	if err != nil {
+		return sa, err
+	}
+	digest, err := utils.Keccak256(b)
+	if err != nil {
+		return sa, err
+	}
+	sa.ID = common.ToHex(digest)
+
+	return sa, nil
+}
+
+// ServiceAgreementRequest represents a service agreement as requested over the wire.
+type ServiceAgreementRequest struct {
+	JobSpec     JobSpec
+	Encumbrance Encumbrance
+	Digest      string
+}
+
+// UnmarshalJSON fulfills Go's built in JSON unmarshaling interface.
+func (sar *ServiceAgreementRequest) UnmarshalJSON(input []byte) error {
+	var jsr JobSpecRequest
+	if err := json.Unmarshal(input, &jsr); err != nil {
+		return err
+	}
+	js, err := NewJobFromRequest(jsr)
+	if err != nil {
+		return err
+	}
+	var en Encumbrance
+	if err := json.Unmarshal(input, &en); err != nil {
+		return err
+	}
+
+	sar.JobSpec = js
+	sar.Digest = js.Digest
```

Comment:
We could definitely simplify this, but I think it might make more sense to get rid of the ServiceAgreementRequest's digest instead of the JobSpec's digest. We refer to the JobSpec digest in the future, but not the SAR.